### PR TITLE
feat: support specPattern, deprecate integrationFolder and componentFolder

### PR DESCRIPTION
**Pull request by lmiller1990**
Documented here (Cypress team only): https://www.notion.so/cypressdx/testFiles-specPattern-glob-patterns-295d545a0af942ebb5b71eda2a6d422b

- deprecate `testFiles`, `integrationFolder`, `component`, superseded by `specPattern`, configurable on `component` and `e2e`

**important** this PR is really big since the concept of `integrationFolder` is sprinkled throughout the entire system. The main, important, and possibly breaking change is explained here: https://github.com/cypress-io/cypress/pull/19319/files#r765721951 please read this comment and ask if you have any concerns. @cowboy is the main person to talk to about this change in behavior.

## Before

```
{
  integrationFolder: 'cypress/integration',
  testFiles: '**/*'
}
```

## After

```js
{
  e2e: {
    specPattern: 'cypress/integration/**/*',
  }
}
```

## Scaffolding `cypress/integration` is gone

Our scaffolding logic is very closely tied to `cypress/integration` existing. Since we no longer scaffold by default on first use, instead using a visual wizard to let the user choose which specs they'd like to scaffold, I deleted all this code and relevant tests.

## `util/spec.ts` is gone

Now using the `findSpecs` method in data-context, which does the same thing with much less indirection, not relying on `integration` or `component` folders any more, instead just using the new `specPattern` property. We only hit the file system once, then caching the specs in `ctx.currentProject.specs`, making it nice and quick.

## Breaking change for `@cypress/code-coverage`

No `integrationFolder` is a breaking change for code coverage. I patched it locally to get CI passing for now: https://github.com/cypress-io/cypress/pull/19319/files#diff-84b549ff6d349070e8dfaf1b1713e4213274611342147dfcc2456848147d72f6. We will need to figure out the migration strategy here. [Made an issue here](https://github.com/cypress-io/code-coverage/issues/536).

## Coming in future PR

- `cypress/integration/**/*` instead of `cypress/e2e/**/*.cy.js`
- update `ignoreTestFiles` to `ignoreSpecPattern`

As outlined [in notion](https://www.notion.so/cypressdx/testFiles-specPattern-glob-patterns-295d545a0af942ebb5b71eda2a6d422b), we are changing the default directory for e2e from `cypress/integration` to `cypress/e2e`. Changing the default means changing **all** our projects in `system-tests/projects` (all 89 of them) and all the ones in `npm`. It's easy but adds a lot of noise to the PR. For this reason, I'll do those in a separate PR ([WIP HERE](https://github.com/cypress-io/cypress/pull/19345)).

### CODE REVIEW

File: packages/data-context/src/actions/ProjectActions.ts [hunk 1]
```diff
@@ -461,8 +468,7 @@ export class ProjectActions {
       throw Error(`Cannot create spec without activeProject.`)
     }
 
-    const config = await this.ctx.project.getConfig(project.projectRoot)
-    const integrationFolder = config.integrationFolder || project.projectRoot
+    const integrationFolder = 'cypress/integration' || project.projectRoot
```
Comment (lmiller1990) [inline packages/data-context/src/actions/ProjectActions.ts:471]: This will need to be updated in another PR, it's around the logic to scaffold specs from the unified runner UI - trying to keep this PR smallish.

File: packages/data-context/src/sources/ProjectDataSource.ts [hunk 2]
```diff
@@ -1,10 +1,75 @@
...
+
+type SpecWithRelativeRoot = FoundSpec & { relativeToCommonRoot: string }
+
+interface MatchedSpecs {
+  projectRoot: string
+  testingType: Cypress.TestingType
+  specAbsolutePaths: string[]
+  specPattern: string | string[]
+}
+export function matchedSpecs ({
```
Comment (lmiller1990) [inline packages/data-context/src/sources/ProjectDataSource.ts:20]: This is the main "logical" change. 

Right now in the current production version of Cypress, if you have:

- `cypress/integration/foo.spec.js`
- `cypress/integration/foo/bar.spec.js`

The videos goes to:
- `cypress/videos/foo.spec.js.mp4`
- `cypress/videos/foo/bar.spec.js.mp4`

Screenshots:
- `cypress/screenshots/foo.spec.js/<screenshot.png>`
- `cypress/screenshots/foo/bar.spec.js/<screenshot.png>`

We always truncate the `cypress/integration` part of the `spec` when saving the artifact. Note it is `cypress/videos/foo.spec.js.mp4`, not `cypress/videos/cypress/integration/foo.spec.js.mp4`. This is easy using the `integrationFolder` property and just truncating it. We can't do this same thing with just a `specPattern`.

To maintain this behavior, ths code here finds the "common" path over all the specs and truncates it. Previously this was always `cypress/integration`. Now it's a bit harder to find.

There is some unsolvable edge cases. See this test: https://github.com/cypress-io/cypress/pull/19319/files#diff-59bf8023d3f96e2d91bc479be5372ebde0317cf6040e5e514ceb276ac71c1342R74-R87 OR https://github.com/cypress-io/cypress/blob/develop/system-tests/test/screenshot_nested_file_spec.js#L7.

Given `cypress/integration/nested-1/nested-2/foo.spec.js, there's no way to replicate the existing behavior - in the current system we'd just remove `cypress/integration`, which is the `integrationFolder`. This is no longer something we can do, since there's no such concept of a "integration folder" - just a glob pattern.

File: circle.yml [hunk 3]
```diff
@@ -1874,46 +1874,6 @@ jobs:
...
-            npm install /root/cypress/cypress.tgz \
-              typescript jest @types/jest enzyme @types/enzyme
-      - run:
-          name: Scaffold and test examples 🏗
-          working_directory: <<parameters.wd>>
-          environment:
-            CYPRESS_INTERNAL_FORCE_SCAFFOLD: "1"
-          command: |
-            echo 'module.exports = {}' > cypress.config.js
-            npx cypress run
```
Comment (lmiller1990) [inline circle.yml:1915]: Todo, put this back in https://cypress-io.atlassian.net/browse/UNIFY-657

File: npm/react/plugins/utils/get-transpile-folders.js [hunk 4]
```diff
@@ -18,7 +18,28 @@ function getTranspileFolders (config) {
     folders.push(config.supportFolder)
   }
 
-  return folders
+  // attempt to add directories based on spec pattern
```
Comment (ZachJW34) [inline npm/react/plugins/utils/get-transpile-folders.js:21]: I think we need to find a better way to accomplish this. For example, I changed the `specPattern` in `npm/react/examples/react-scripts/cypress.config.js` to `'**/*cy-spec.js'` and the result was `dirsFromSpecPattern = '/Users/zachjw/work/cypress/npm/react/examples/react-scripts/**/*cy-spec.j'`

I'm trying to think of a more proper way to handle this but it's a pretty difficult problem. Could we use globby on the spec pattern and extract from the returned files/dirs? That approach might fall short if they haven't added a spec yet or they add a new one that results in a new folder...

Also, there is a check on line 9 that looks like it can be removed i.e. `if (config.componentFolder)`

File: packages/app/cypress.config.ts [hunk 5]
```diff
@@ -11,19 +11,15 @@ export default defineConfig({
     runMode: 2,
     openMode: 0,
   },
-
-  testFiles: '**/*.{spec,cy}.{js,ts,tsx,jsx}',
+  nodeVersion: 'system',
```
Comment (ZachJW34) [inline packages/app/cypress.config.ts:14]: This should be removed here and in other configs that this got added to.

File: packages/data-context/src/sources/ProjectDataSource.ts [hunk 6]
```diff
@@ -1,10 +1,75 @@
...
+  }
+
+  const specs = specAbsolutePaths.map((absolute) => {
+    return transformSpec(projectRoot, absolute, testingType, commonRoot)
+  })
+
+  return specs
+}
+
+export function transformSpec (projectRoot: string, absolute: string, testingType: Cypress.TestingType, commonRoot: string): SpecWithRelativeRoot {
```
Comment (ZachJW34) [inline packages/data-context/src/sources/ProjectDataSource.ts:42]: We have a similar function in `FileDataSource` called `normalizeFileToSpec`.  In the future, I think we should have one place where we do this normalizing so something to keep in mind

File: packages/graphql/src/schemaTypes/objectTypes/gql-CurrentProject.ts [hunk 7]
```diff
@@ -75,9 +74,9 @@ export const CurrentProject = objectType({
 
     t.connectionField('specs', {
       description: 'Specs for a project conforming to Relay Connection specification',
-      type: Spec,
-      nodes: (source, args, ctx) => {
-        return ctx.project.findSpecs(source.projectRoot, ctx.appData.currentTestingType === 'component' ? 'component' : 'integration')
+      type: 'Spec',
+      nodes: async (source, args, ctx) => {
```
Comment (ZachJW34) [inline packages/graphql/src/schemaTypes/objectTypes/gql-CurrentProject.ts:78]: `async` is not necessary

File: packages/server/lib/modes/run.js [hunk 8]
```diff
@@ -1535,7 +1539,7 @@ module.exports = {
       options.browsers = browsers
 
       return createAndOpenProject(socketId, options)
-      .then(({ project, projectId, config, configFile }) => {
+      .then(async ({ project, projectId, config, configFile }) => {
```
Comment (ZachJW34) [inline packages/server/lib/modes/run.js:1542]: `async` is not necessary

File: cli/types/cypress.d.ts [hunk 9]
```diff
@@ -2548,8 +2548,14 @@ declare namespace Cypress {
...
      * A String or Array of glob patterns used to ignore test files that would otherwise be shown in your list of tests. Cypress uses minimatch with the options: {dot: true, matchBase: true}. We suggest using http://globtester.com to test what files would match.
      * @default "*.hot-update.js"
+     * @deprecated use `ignoreSpecPattern` instead
      */
     ignoreTestFiles: string | string[]
+    /**
+     * A String or Array of glob patterns used to ignore test files that would otherwise be shown in your list of tests. Cypress uses minimatch with the options: {dot: true, matchBase: true}. We suggest using http://globtester.com to test what files would match.
+     * @default "*.hot-update.js"
+     */
+    ignoreSpecPattern: string | string[]
```
Comment (ZachJW34) [inline cli/types/cypress.d.ts:2558]: Is the implementation of this going to be a separate PR?

File: npm/angular/patches/@cypress+code-coverage+3.9.5.patch [hunk 10]
```diff
@@ -0,0 +1,45 @@
+diff --git a/node_modules/@cypress/code-coverage/support-utils.js b/node_modules/@cypress/code-coverage/support-utils.js
```
Comment (ZachJW34) [inline npm/angular/patches/@cypress+code-coverage+3.9.5.patch:1]: We should create an issue for updating `@cypress/code-coverage`

**Review by ZachJW34 — COMMENTED** [review]
Just some small things. Awesome work, I tested as well as I could and it works great!

File: cli/types/cypress.d.ts (hunk 9 above)
Comment (lmiller1990) [reply cli/types/cypress.d.ts:2558]: Should already be done - it's just a find+replace from `ignoreTestPattern` -> `ignoreSpecPattern`.

File: npm/angular/patches/@cypress+code-coverage+3.9.5.patch (hunk 10 above)
Comment (lmiller1990) [reply npm/angular/patches/@cypress+code-coverage+3.9.5.patch:1]: Yep we have it https://github.com/cypress-io/code-coverage/issues/536

File: packages/graphql/src/schemaTypes/objectTypes/gql-CurrentProject.ts (hunk 7 above)
Comment (lmiller1990) [reply packages/graphql/src/schemaTypes/objectTypes/gql-CurrentProject.ts]: https://github.com/cypress-io/cypress/pull/19319/commits/9a13720f8d4c881a8edac8408c9239a353fc8ac5

File: packages/app/cypress.config.ts (hunk 5 above)
Comment (lmiller1990) [reply packages/app/cypress.config.ts]: https://github.com/cypress-io/cypress/pull/19319/commits/9a13720f8d4c881a8edac8408c9239a353fc8ac5

File: npm/react/plugins/utils/get-transpile-folders.js (hunk 4 above)
Comment (lmiller1990) [reply npm/react/plugins/utils/get-transpile-folders.js:16]: Removed componentFolder https://github.com/cypress-io/cypress/pull/19319/commits/9a13720f8d4c881a8edac8408c9239a353fc8ac5

Not sure on how to handle transpile folders for a spec glob... I wish CRA didn't have this silly restriction in the first place. Your globby solution sounds interesting, but not sure if this will impact perf (adding dozens, if not hundreds of directories to the watcher).

File: cli/types/cypress.d.ts (hunk 9 above)
Comment (lmiller1990) [reply cli/types/cypress.d.ts:2558]: The pr is wip and here https://github.com/cypress-io/cypress/pull/19345

File: packages/data-context/src/sources/ProjectDataSource.ts (hunk 6 above)
Comment (lmiller1990) [reply packages/data-context/src/sources/ProjectDataSource.ts:42]: Yes definitely, will look at cleaning this up soon

File: npm/react/plugins/utils/get-transpile-folders.js (hunk 4 above)
Comment (ZachJW34) [reply npm/react/plugins/utils/get-transpile-folders.js:16]: I wonder if we can just add the `cypress` folder and call it a day.

File: npm/react/examples/webpack-file/cypress.config.js [hunk 11]
```diff
@@ -1,10 +1,10 @@
 module.exports = {
   'video': false,
   'fixturesFolder': false,
-  'testFiles': '**/*cy-spec.js',
   'viewportWidth': 500,
   'viewportHeight': 500,
   'component': {
+    'specPattern': '**/*cy-spec.js',
```
Comment (ryanthemanuel) [inline npm/react/examples/webpack-file/cypress.config.js:7]: Should this be cypress/component/**/*cy-spec.js?  Seems like we did that in npm/react/examples/webpack-options/cypress.config.js

File: cli/types/cypress.d.ts [hunk 12]
```diff
@@ -2780,6 +2788,12 @@ declare namespace Cypress {
     /**
      * Glob pattern to determine what test files to load.
      */
+    specPattern: string | string[]
```
Comment (tbiethman) [inline cli/types/cypress.d.ts:2791]: Theres a note in the Notion doc regarding specPattern usage:
> specPattern may only be specified in the e2e and component objects, and is invalid at the top level of the config (will error if specified at the top level)

Should `specPattern` be defined for this type if its usage is invalid at the top level? Or is it a pain to compose it into the `e2e`/`component` config types individually?

File: packages/data-context/src/sources/ProjectDataSource.ts (hunk 2 above)
Comment (ryanthemanuel) [reply packages/data-context/src/sources/ProjectDataSource.ts:20]: Will this get weird as teams add tests over time?  Something like this:

*First Release*

File locations
* cypress/integration/a/b/c.spec.js
Screenshot locations
* cypress/screenshots/c.spec.js/<screenshot.png>

*Second Release*

File locations
* cypress/integration/a/b/c.spec.js
* cypress/integration/a/d/e.spec.js
Screenshot locations
* cypress/screenshots/b/c.spec.js/<screenshot.png>
* cypress/screenshots/d/e.spec.js/<screenshot.png>

So in the first release, screenshots will be in one location, but in a subsequent release they'll be in a different location (even though the file location hasn't changed).

Do we foresee that causing any issues?

File: packages/server/lib/modes/run.js [hunk 13]
```diff
@@ -119,7 +126,15 @@ const formatSymbolSummary = (failures) => {
   return getSymbol(failures)
 }
 
-const formatPath = (name, n, colour = 'reset') => {
+const macOSRemovePrivate = (str) => {
+  if (str.startsWith('/private')) {
```
Comment (ryanthemanuel) [inline packages/server/lib/modes/run.js:130]: Is it possible that we're not on a mac and legitimately have a directory that starts with private?

File: npm/react/examples/webpack-file/cypress.config.js (hunk 11 above)
Comment (lmiller1990) [reply npm/react/examples/webpack-file/cypress.config.js]: Yep, fixed https://github.com/cypress-io/cypress/pull/19319/commits/dc67b143ed92ac72a306d437dc83a236aa69fb81#diff-d9606627b12888cd15239859eb4597600250d590fafbf51236413a5633a89b52L7

File: packages/server/lib/modes/run.js (hunk 13 above)
Comment (lmiller1990) [reply packages/server/lib/modes/run.js]: Sure https://github.com/cypress-io/cypress/pull/19319/commits/dc67b143ed92ac72a306d437dc83a236aa69fb81#diff-58fd4a8e28c13b52b19c42346eade66ed1de5a8b139541dff4e94fef42c3c701R131

Note: this is only used for the system tests; it won't impact general users, it's only called [here](https://github.com/cypress-io/cypress/pull/19319/commits/dc67b143ed92ac72a306d437dc83a236aa69fb81#diff-58fd4a8e28c13b52b19c42346eade66ed1de5a8b139541dff4e94fef42c3c701R143). Either way, I added an additional check.

File: cli/types/cypress.d.ts (hunk 12 above)
Comment (lmiller1990) [reply cli/types/cypress.d.ts:2791]: Top level `specPattern` should be invalid. For now, we don't have a good way to validate "only top level" or "only nested" config yet, but that will be done soon, partially in [this PR](https://github.com/cypress-io/cypress/pull/19347) and the PR that'll follow it up. I'll add you as a reviewer when this is updated; for now, to minimize noise and PR size, there isn't any additional validation implemented in this PR.

File: packages/runner/cypress/integration/runner.ui.spec.js [hunk 14]
```diff
@@ -405,43 +405,4 @@ describe('src/cypress/runner', () => {
       cy.get('.reporter').should('not.exist')
     })
   })
-
-  describe('inline spec list', () => {
```
Comment (chrisbreiding) [inline packages/runner/cypress/integration/runner.ui.spec.js:409]: Why was this test removed?

File: packages/server/lib/project-base.ts [hunk 15]
```diff
@@ -446,25 +425,17 @@ export class ProjectBase<TServer extends Server> extends EE {
...
   }: {
-    specs: Cypress.Cypress['spec'][]
+    specs: FoundSpec[]
     config: Cfg
   }) {
-    const specsStore = new SpecsStore(config, this.testingType)
+    const specsStore = new SpecsStore()
 
     const startSpecWatcher = () => {
-      return specsStore.watch({
```
Comment (chrisbreiding) [inline packages/server/lib/project-base.ts:455]: How are specs watched and the specs list updated without starting a watcher?

File: packages/data-context/test/unit/sources/ProjectDataSource.spec.ts [hunk 16]
```diff
@@ -0,0 +1,77 @@
+import { matchedSpecs } from '../../../src/sources'
+import { expect } from 'chai'
+
+describe('ProjectDataSource', () => {
```
Comment (chrisbreiding) [inline packages/data-context/test/unit/sources/ProjectDataSource.spec.ts:4]: I think the test coverage should be expanded to cover more of what `transformSpec` does. For example, the handling of backslashes (`\`) in paths and the various spec files extensions.

File: packages/data-context/src/actions/ProjectActions.ts [hunk 17]
```diff
@@ -495,6 +501,15 @@ export class ProjectActions {
       }
     })
 
+    const specPattern = await this.ctx.project.specPatternForTestingType(project.projectRoot, 'e2e')
+
+    if (!specPattern) {
```
Comment (chrisbreiding) [inline packages/data-context/src/actions/ProjectActions.ts:506]: I'm guessing this can happen if the user has configured it to an invalid value? Seems like it should be handled by config validation instead.

**Review by chrisbreiding — CHANGES_REQUESTED** [review]

File: packages/runner/cypress/integration/runner.ui.spec.js (hunk 14 above)
Comment (lmiller1990) [reply packages/runner/cypress/integration/runner.ui.spec.js:409]: I don't think this feature or spec makes sense anymore. Basically, this was testing the ability to have inline spec list in the e2e runner. With how `specPattern`is implemented (similar to `--spec`), we only glob for the specs we want - in this case, when executing `runner.ui.spec.js` in run mode, that would be this spec file.

Previously, even when `--spec` was provided, we still globbed all the `testFiles`. Since we only glob for the relevant spec files now, in run mode, the inline spec list only shows 1 spec file, meaning you cannot switch between specs using the UI in run mode (which doesn't make much sense anyway).

Here's an image. We will not show the spec list during run mode (and previously we shouldn't have shown in anyway) so this test didn't make a ton of sense in the first place, and even less sense now.

image

Comment (lmiller1990) [reply packages/runner/cypress/integration/runner.ui.spec.js:409]: I made sure to remove the code relevant to this `CypressInternal_UseInlineSpecList` too: https://github.com/cypress-io/cypress/pull/19319/commits/f326e88500fce17fe84dbf40be4f67cb5d220a81

File: packages/server/lib/project-base.ts (hunk 15 above)
Comment (lmiller1990) [reply packages/server/lib/project-base.ts:455]: This is moving out of `project-base.ts` into the data context. I don't know if this is implemented yet.

This change does break `runner-ct`'s inline spec list, but that's okay, since we are well and truly using the unified runner ass the main runner now. I removed most of the spec store code, since we no longer need this - we are serving the specs via GraphQL now, not the spec store.

File: packages/data-context/src/actions/ProjectActions.ts (hunk 17 above)
Comment (lmiller1990) [reply packages/data-context/src/actions/ProjectActions.ts:506]: Yep, I just put this here to appease TypeScript. We need to update the configuration validation a bunch, at this point in the code we should have validating everything already. A lot of work around config is [happening here](https://github.com/cypress-io/cypress/pull/19347) and in another PR. I didn't add any validation here other than some simple inline checks in order to keep this PR "small" (if you can really call 100 files small...)

File: packages/data-context/test/unit/sources/ProjectDataSource.spec.ts (hunk 16 above)
Comment (lmiller1990) [reply packages/data-context/test/unit/sources/ProjectDataSource.spec.ts]: I added some more tests: https://github.com/cypress-io/cypress/pull/19319/commits/6cc7047e23157a26e4f9b1e14b1198c9bc1b4313

We seem to do a lot of checks around `\` all over the place, definitely want to consolidate this. Eg, it's not clear if we can rely on `projectRoot` to be posixified at this point, so for good measure I am normalizing the `\` to `/` here now.

**Review by chrisbreiding — DISMISSED** [review]

**Review by ryanthemanuel — DISMISSED** [review]
